src/go/k8s/internal/controller/pvcunbinder/pvcunbinder.go

```diff
@@ -0,0 +1,314 @@
+// the lack of transactions in Kubernetes/etc and the need to operate across
+// many objects, it's quite difficult to guarantee this. The general strategy
+// is to fetch a snapshot of the world as early as possible and then rely on
+// ResourceVersions to inform us about changes from external actors, in which
+// case we'll re-queue.
+// TODO use an in memory timeout to prevent complete unbinding of Pods.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := ctrl.LoggerFrom(ctx).WithName("PVCUnbinder")
+	ctx = log.IntoContext(ctx, logger)
+
+	var pod corev1.Pod
+	if err := r.Client.Get(ctx, req.NamespacedName, &pod); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if ok, requeueAfter := r.shouldRemediate(ctx, &pod); !ok || requeueAfter > 0 {
+		logger.Info("shouldn't remediate Pod; skipping", "name", pod.Name, "ok", ok, "requeue-after", requeueAfter)
+		return ctrl.Result{RequeueAfter: requeueAfter, Requeue: ok}, nil
+	}
+
+	// NB: We denote PVCs that are deleted as a nil entry within this map. If a
+	// PVC is not to be considered, it should be removed from this map.
+	pvcByKey := map[client.ObjectKey]*corev1.PersistentVolumeClaim{}
+
+	for _, pvcKey := range StsPVCs(&pod) {
+		var pvc corev1.PersistentVolumeClaim
+		if err := r.Client.Get(ctx, pvcKey, &pvc); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return ctrl.Result{}, err
+			}
+			pvcByKey[pvcKey] = nil
+			continue
+		}
+		pvcByKey[pvcKey] = &pvc
+	}
+
+	// If there are no StatefulSet managed PVCs, there's nothing we can do.
+	if len(pvcByKey) == 0 {
+		logger.Info("Pod had no detectable StatefulSet PVCs. Skipping.")
+		return ctrl.Result{}, nil
+	}
+
+	// Nothing can be done to scope this query unless we decide to bind the
+	// implementation to rancher's local path provisioner which adds a label we
+	// could query against.
+	var pvList corev1.PersistentVolumeList
+	if err := r.Client.List(ctx, &pvList); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Filter PVs down to ones that are:
+	// 1. Bound to a PVC we care about.
+	// 2. Have a NodeAffinity (which we assume is the cause of our Pod being in Pending)
+	var pvs []*corev1.PersistentVolume
+	for i := range pvList.Items {
+		pv := &pvList.Items[i]
+
+		if pv.Spec.ClaimRef == nil {
+			continue
+		}
+
+		key := client.ObjectKey{
+			Name:      pv.Spec.ClaimRef.Name,
+			Namespace: pv.Spec.ClaimRef.Namespace,
+		}
+
+		// Skip over any PVs that aren't bound to one of our targeted PVCs
+		_, ok := pvcByKey[key]
+		if !ok {
+			continue
+		}
+
+		// Filter out PVCs and PVs that don't have a NodeAffinity or aren't a
+		// HostPath node. We can't be certain about to handle those.
+		if pv.Spec.NodeAffinity == nil || pv.Spec.HostPath == nil {
+			delete(pvcByKey, key)
+			continue
+		}
+
+		pvs = append(pvs, pv)
+	}
+
+	// 3. Ensure that all PVs have reclaim set to Retain
```

Comment:
I really like that you've numbered the major steps in the process here. It helps me know when to clear out my mental "cache" of relevant code I've read.